src/microbe_stage/BiomeConditions.cs

```diff
@@ -170,6 +172,124 @@ public bool TryGetCompound(Compound compound, CompoundAmountType amountType, out
+    /// </param>
+    /// <param name="changes">
+    ///   The changes to apply. Changes that would negative compounds are clamped automatically
+    /// </param>
+    /// <param name="newCloudSizes">
+    ///   Specifies the compound cloud spawn sizes for all new non-environmental compounds
+    /// </param>
+    public void ApplyLongTermCompoundChanges(Biome biomeDetails, Dictionary<Compound, float> changes,
+        IReadOnlyDictionary<Compound, float> newCloudSizes)
+    {
+        var simulationParameters = SimulationParameters.Instance;
+
+        // Apply first non-environmental types as that is the easiest
+        foreach (var entry in changes)
+        {
+            var definition = simulationParameters.GetCompoundDefinition(entry.Key);
+            if (!definition.IsEnvironmental)
+            {
+                if (!TryGetCompound(entry.Key, CompoundAmountType.Biome, out var existing))
+                {
+                    if (!newCloudSizes.TryGetValue(entry.Key, out var cloudSize))
+                    {
+                        GD.PrintErr(
+                            $"Unknown cloud spawn size to use for {entry.Key}, using a default hardcoded value");
+                        cloudSize = 250000;
+                    }
+
+                    existing.Amount = cloudSize;
+                }
+
+                existing.Density += entry.Value;
+                ModifyLongTermCondition(entry.Key, existing);
+            }
+            else if (!definition.IsGas)
+            {
+                // Then non-gas environmental can also be applied here as these don't use custom handling
+                TryGetCompound(entry.Key, CompoundAmountType.Biome, out var existing);
+
+                existing.Ambient = Math.Clamp(existing.Ambient + entry.Value, 0, 1);
+                ModifyLongTermCondition(entry.Key, existing);
+            }
+        }
+
+        // Then apply environmental all at once as absolute values so that the gas percentages will not add up to over
+        // 100%
+
+        // Calculate current gases in absolute volume
+        var gases = new Dictionary<Compound, float>();
+        float previousTotal = 0;
+
+        foreach (var current in compounds)
+        {
+            if (simulationParameters.GetCompoundDefinition(current.Key).IsGas)
+            {
+                var absoluteAmount = biomeDetails.GasVolume * current.Value.Ambient;
+                gases[current.Key] = absoluteAmount;
+                previousTotal += absoluteAmount;
+            }
+        }
+
+        float previousOther = 1 * biomeDetails.GasVolume - previousTotal;
+
+        // TODO: handling for keeping some space for other compounds?
+
+        foreach (var change in changes)
+        {
+            if (!simulationParameters.GetCompoundDefinition(change.Key).IsGas)
+                continue;
+
+            gases.TryGetValue(change.Key, out var updatedValue);
+
+            updatedValue = Math.Clamp(updatedValue + change.Value, 0, 1);
```

Comment:
Shouldn't the maximum value for the clamp be `biomeDetails.GasVolume`? Otherwise, with a high enough gas volume, the `updatedValue` may be higher than 1 and cause a clamp when it shouldn't.

Reply:
Yeah, I think that is true on second read of that code so I'll make that change. Good catch.